qutebrowser/mainwindow/tabwidget.py

```diff
@@ -329,6 +335,18 @@ def update_tab_favicon(self, tab: QWidget):
             if config.val.tabs.tabs_are_windows:
                 self.window().setWindowIcon(self.window().windowIcon())
 
+        if tab.data.pinned:
+            self.tabBar().clear_pinned_statistics.emit()
+
+    def removeTab(self, idx):
+        """Additions to removeTab."""
+        # We must clear cache BEFORE the tab is removed, as the removeTab
+        # function calls tabSizeHint to resize the tabs after removing the tab.
+        # If we assume super().removeTab will NOT call tabSizeHint prematurely,
+        # this should work.
+        self.tabBar().clear_pinned_statistics.emit()
```

Comment:
I really don't like how sensitive this is to the exact order things are done in, so I would like to find a better way to invalidate the cache if possible. One way that I thought of is that we can keep a count of tabs independent from the QTabBar, and when they differ invalidate the cache (in tabSizeHint). This has the advantage of being able to more 'correctly' handle additions/removals if the tabs are resized (which they should be, by qt), but if a new tab is added before the redraw there can be problems.

In order to prevent a crash when this happens I clamped the division by zero below, not a great solution, but I'd much rather have mis-sized tabs than a crash if we mess up this logic.

I thought about this a bit and we might not actually need any of this (unless I forgot a feature with tab sizing). I'll submit a PR and hopefully we can discard this one...